### ISSUE TITLE
Primitive functions on ShardedDeviceArray

### ISSUE THREAD

**Issue by awav** [google/jax#2535]
Hello JAX team,

I was surprised to see that `ShardedDeviceArray` collapses to a standard shaped array after applying a primitive function to it. Would it make sense to keep `ShardedDeviceArray` type as an output for functions like `exp` with the same type inputs? I think it would help to write clean and modular code.

Thanks

**Comment by mattjj** [comment]
The current policy is that only `pmapped` functions will maintain sharding across devices. That usually means you can just apply `pmap(np.exp)` instead of `np.exp` to maintain sharding. It's pretty explicit and predictable.

It's true that at least in some cases, particularly for unary primitives for which the semantics is already to map over their array arguments, we could instead "do the right thing" and execute the pmapped version automatically.

It gets a little tricky, though: what about binary primitives like `+`? If the two inputs are ShardedDeviceArrays with equal shapes and equivalent shardings then we could maintain that for the output. But what about broadcasting semantics? What if one input is a host (cpu) scalar? What about primitives that don't have mapped semantics, like `dot`?

That is, I think the trickiest bit here would be deciding on a heuristic policy for each primitive (though they may cluster into a few groups, like mapped unary/binary/ternary, contractions, etc) and then implementing that policy. It's not obvious a priori that a system of these heuristics will have a good power-to-implementation-weight ratio, or that they will lead to predictable behavior in all cases.

For those reasons, this seems like a low priority feature. Something that could bump up the priority would be some compelling examples of how it helps write clean and modular code, where just calling `pmap` explicitly doesn't cut it for some reason.

WDYT?

**Comment by awav** [comment]
I agree with most of the points that you have listed. It would be pretty easy to implement unary functions for sharded arrays, and it becomes messy for functions with more than one argument. However, for binary primitives, the sharding axis could be considered as a leading dimension.

> Something that could bump up the priority would be some compelling examples of how it helps write clean and modular code, where just calling pmap explicitly doesn't cut it for some reason.

I don't have a particular example at the moment, but let's say we have a generic algorithm, that uses a function A that is replaceable, and someone could implement a pmapped version of A that spits out a sharded array for memory-wise reasons, that would be amazing if the rest of algorithm could still work on shards only.

**Comment by mrahtz** [comment]
+1 - we've just been bitten by this.

Our use-case is that we're dealing with pytrees representing the activations of a large sharded model. We want to be able to do various manipulations on those activations - for example, allocating a similarly-shaped pytree of zeros as a mask, using something like `jax.tree_map(jnp.zeros_like, activations)`. This isn't happening as part of a training loop, it's happening after the model is trained and we're analysing it - so putting it inside a `pmap` would unnecessarily complicate our code.

I haven't thought about it too hard, but it seems like the following policy might suffice?
* Unary operators: preserve sharding
* Binary operators: preserve sharding if the two arguments are sharded exactly the same way, else throw an error (since it's better to expose this non-obvious behaviour than to silently just convert everything to a `DeviceArray`)

I can imagine the implementation to make this possible might well be difficult, though, so given the fact that it's only the two of us requesting this change, I accept that the tradeoff probably isn't worth it.

**Comment by hawkinsp** [comment]
This issue is stale, since `Array` replaced `ShardedDeviceArray`, and we support much richer policies about sharding propagation.
